### ISSUE TITLE
rspec failing with undefined role_names with legacy_connection_handling = false in Rails 7

### ISSUE THREAD

**Issue by mmpeters75** [rails/rails#47833]
### Steps to reproduce
I set config.active_record.legacy_connection_handling = false in my application. I have only one database. The app is running fine (with puma) but I have config.use_transactional_fixtures = true set in my rails_helper.rb file for rspec, and setting that to true causes the following error:

```NoMethodError: undefined method `role_names' for #<ActiveRecord::ConnectionAdapters::LegacyPoolManager```

I could probably resolve this by setting use_transactional_fixtures to false and using database cleaner instead but this should work, either by some config I'm clearly missing or something else. 

database.yml

```test:
  min_messages: WARNING
  adapter: postgresql
  encoding: unicode
  host: <%= ENV['DATABASE_HOST'] || 'localhost' %>
  database: <%= ENV['DATABASE_NAME'] %><%= ENV['TEST_ENV_NUMBER'] %>
  pool: 5
  username: <%= ENV['DATABASE_UID'] %>
  password: <%= ENV['DATABASE_PWD'] %>
  template: template0
  checkout_timeout: 5
  connect_timeout: 120
  read_timeout: 30
```

### Expected behavior
Rspec test runs

### Actual behavior
```NoMethodError: undefined method `role_names' for #<ActiveRecord::ConnectionAdapters::LegacyPoolManager:0x00007fd8410e0a58 @name_to_pool_config={:default=>#<ActiveRecord::ConnectionAdapters::PoolConfig:0x00007fd8430c0008 @_mutex=#<Thread::Mutex:0x00007fd8430cbef8>, @connection_class=ActiveRecord::Base, @db_config=#<ActiveRecord::DatabaseConfigurations::HashConfig:0x00007fd8430c14f8 @env_name="test", @name="primary", @configuration_hash={:min_messages=>"WARNING", :adapter=>"postgresql", :encoding=>"unicode", :host=>"localhost", :database=>"spring-web_test", :pool=>5, :username=>"spring_web", :password=>"devpassword", :template=>"template0", :checkout_timeout=>5, :connect_timeout=>120, :read_timeout=>30}>, @role=:writing, @shard=:default, @pool=#<ActiveRecord::ConnectionAdapters::ConnectionPool:0x00007fd8430cb1b0 @mon_data=#<Monitor:0x00007fd8430cb070>, @mon_data_owner_object_id=20520, @query_cache_enabled=#<Concurrent::Map:0x00007fd8430cafa8 entries=0 default_proc=#<Proc:0x00007fd8430caeb8 /Users/matt.peters/git/klarna/spring-web/vendor/cache/ruby/3.0.0/gems/activerecord-7.0.4/lib/active_record/connection_adapters/abstract/query_cache.rb:32>>```

### System configuration
Rails version
7.0.4.

**Ruby version**:
3.0.3

**Comment by eileencodes** [comment]
Hi @mmpeters75 thanks for the report. Can you make a new app that reproduces the bug? Without a reproduction we can't do much to debug this as we don't know what the requirements are to reproduce.

**Comment by davinlagerroos** [comment]
In case other people run across this, we ran in to a similar issue when we upgraded to rails 6.1 and opted in to the new defaults. We were getting this error when running `bin/rails test:system` or `bin/rails test:all`. Running `bin/rails test` or `bin/rails test <some/test/file>` worked fine.

You might see if you have something in your app that establishes an active record connection that could be running before rails gets fully loaded. In our case we had a custom rake task. The flow seems to have been:
* run `bin/rails test:system`
* Rakefile loads
* the Rakefile loaded the configuration for the custom task
* the custom task configuration called `ActiveRecord::Base.establish_connection`. Without all of rails loaded, the value of `ActiveRecord.legacy_connection_handling` was the fallback value of `true` so the `LegacyPoolManager` was used
* The connection handler gets stored in `Thread`
* the rest of rails loads, including the configuration for `active_record.legacy_connection_handling`
* active record test setup methods run. ActiveRecord looks at `ActiveRecord.legacy_connection_handling` which is now set to `false` so active record thinks it should be talking to `PoolManager`, but when it fetches the connection handler it gets the legacy pool manager from before, which does not respond to `role_names`

In our case the custom task had been imported from another app that only used active record, so it was doing a bunch of unnecessary setup (including the `ActiveRecord::Base.establish_connection`) that we were able to remove
